server/xcode/xcode_darwin.go

```diff
@@ -151,6 +184,48 @@ func xcodePlistForPath(path string) (*xcodePlist, error) {
 	return &xcodePlist, nil
 }
 
+func commandLineToolsVersion(developerDir string) *xcodeVersion {
+	sdkPaths, err := fs.Glob(os.DirFS(developerDir), "SDKs/*")
+	if err != nil {
+		log.Warningf("Error reading CommandLineTools SDKs from %s: %s", developerDir, err.Error())
+		return nil
+	}
+	sdks := make(map[string]string)
+	for _, sdkPath := range sdkPaths {
+		sdks[strings.TrimSuffix(filepath.Base(sdkPath), ".sdk")] = sdkPath
+	}
+	return &xcodeVersion{
+		version:          defaultXcodeVersion,
+		developerDirPath: developerDir,
+		sdks:             sdks,
+	}
+}
+
+// Returns the version of the Xcode Command Line Tools, or "unknown" if it
+// can't be determined.
+func commandLineToolsVersionString() string {
+	out, err := exec.Command(
+		"sh", "-c",
+		"pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '/version:/ {print $2}'",
+	).Output()
+	if err != nil {
+		log.Warningf("Error determining Xcode Command Line Tools version: %s", err.Error())
+		return "unknown"
+	}
+	return strings.TrimSpace(string(out))
+}
+
+// Returns the active developer directory according to xcode-select, or the
+// empty string if xcode-select failed to run.
+func xcodeSelectDeveloperDir() string {
+	out, err := exec.Command("xcode-select", "-p").Output()
+	if err != nil {
+		log.Warningf("Error calling xcode-select: %s", err.Error())
```

Comment:
Should I return an error here instead and let the caller log?

Reply:
This one matters more, but since you are comparing the returned string against some known good values and looking for a match, if you return an empty string here on failure it seems fine -- you can log an error here if you get one and your calling code will just not find a match the returned empty string against anything.